File: src/lib/ssl-errors.ts
import type { DatabaseLocation } from '$lib/connection-presets';

export type ConnectionErrorKind =
	| 'untrusted_ca'
	| 'hostname_mismatch'
	| 'certificate_required'
	| 'unknown';

export type ConnectionErrorClassification = {
	kind: ConnectionErrorKind;
	showTroubleshooting: boolean;
};

export function classifyConnectionError(
	message: string,
	context: { location: DatabaseLocation | null; remoteHost: boolean }
): ConnectionErrorClassification {
	const normalized = message.toLowerCase();

	if (
		normalized.includes('certificate verify failed') ||
		normalized.includes('unknown issuer') ||
		normalized.includes('unknown ca') ||
		normalized.includes('self signed') ||
		normalized.includes('self-signed') ||
		normalized.includes('invalid certificate') ||
		normalized.includes('certificate required')
	) {
		return { kind: 'untrusted_ca', showTroubleshooting: true };
	}

	if (
		normalized.includes('hostname mismatch') ||
		normalized.includes('name mismatch') ||
		normalized.includes('not valid for') ||
		normalized.includes('certificate is not valid for')
	) {
		return { kind: 'hostname_mismatch', showTroubleshooting: true };
	}

	if (
		normalized.includes('wallet') ||
		normalized.includes('tcps') ||
		normalized.includes('ssl required') ||
		normalized.includes('requires ssl') ||
		normalized.includes('requires tls')
	) {
		return { kind: 'certificate_required', showTroubleshooting: true };
	}

	return {
		kind: 'unknown',
		showTroubleshooting: context.location === 'organization' && context.remoteHost
	};
}

File: compose-app/shared/src/main/kotlin/com/safedb/connection/SslErrors.kt
package com.safedb.connection

enum class ConnectionErrorKind {
    UntrustedCa,
    HostnameMismatch,
    CertificateRequired,
    Unknown,
}

data class ConnectionErrorClassification(
    val kind: ConnectionErrorKind,
    val showTroubleshooting: Boolean,
)

fun classifyConnectionError(
    message: String,
    context: ConnectionErrorContext,
): ConnectionErrorClassification {
    val normalized = message.lowercase()
    val hasCertificateHostnameMismatch =
        normalized.contains("certificate is not valid for") ||
            normalized.contains("cert is not valid for") ||
            (normalized.contains("x509") && normalized.contains("not valid for"))

    if (
        normalized.contains("hostname mismatch") ||
        normalized.contains("name mismatch") ||
        hasCertificateHostnameMismatch
    ) {
        return ConnectionErrorClassification(ConnectionErrorKind.HostnameMismatch, showTroubleshooting = true)
    }

    if (
        normalized.contains("wallet") ||
        normalized.contains("tcps") ||
        normalized.contains("ssl required") ||
        normalized.contains("requires ssl") ||
        normalized.contains("requires tls") ||
        normalized.contains("certificate required")
    ) {
        return ConnectionErrorClassification(ConnectionErrorKind.CertificateRequired, showTroubleshooting = true)
    }

    if (
        normalized.contains("certificate verify failed") ||
        normalized.contains("unknown issuer") ||
        normalized.contains("unknown ca") ||
        normalized.contains("self signed") ||
        normalized.contains("self-signed") ||
        normalized.contains("invalid certificate")
    ) {
        return ConnectionErrorClassification(ConnectionErrorKind.UntrustedCa, showTroubleshooting = true)
    }

    return ConnectionErrorClassification(
        kind = ConnectionErrorKind.Unknown,
        showTroubleshooting = context.location == DatabaseLocation.Organization && context.remoteHost,
    )
}

data class ConnectionErrorContext(
    val location: DatabaseLocation?,
    val remoteHost: Boolean,
)
